release-22.1: server: new HTTP API to execute SQL statements

dhartunian opened the pull request:
Backport 2/2 commits from #79663 and #84386.

/cc @cockroachdb/release

---

Requested by @andreimatei .

This commit introduces a new API endpoint `/api/v2/sql/`,
which can execute arbitrary SQL statements over HTTP.

The endpoint is authenticated, i.e. the client must provide
an appropriate `X-Cockroach-API-Session:` header on each request.


There is a Swagger spec to enable auto-documentation of the API.

The request payload should be submitted using JSON.

Summary of request fields:
- `database`: the current database, defaults to `defaultdb`
- `application_name`
- `timeout`: max execution time, defaults to 5s
- `max_results_size`: max size of results, defaults to ~10KB.
- `statements`: which statements to run. An array of {"sql",
  "arguments"} objects. The number of items in "arguments"
  must match the number of placehoders in "sql".
- `execute`: must be set to true to actually execute the SQL.
  Otherwise, the request is merely parsed to verify the syntax is correct.

Example use:

1. Generate some test data and create a user with access to it:

   ```
   $ cockroach gen example-data intro | cockroach sql --certs-dir=...
   $ cockroach sql --certs-dir=... \
       -e 'create user myuser with password abc'
       -e 'grant select on table intro.mytable to myuser'
   ```

2. Generate an HTTP API key, for example using the CLI (an HTTP
   request to /api/v2/login is also possible):

   ```
   $ cockroach session-auth login myuser  --certs-dir=certs
   ```

   (this prints out the session key, assign it manually to the $APIKEY env var.)

3. Run some HTTP API queries from the shell using `curl`. For example:

   ```
   $ curl -k -H "X-Cockroach-API-Session: $APIKEY" \
       -H "Content-Type: application/json" \
       --data-binary '{"database":"intro",
       "statements":[{"sql":"SELECT v FROM mytable WHERE l%2 = 0"}],
       "execute": true}' \
       https://127.0.0.1:8080/api/v2/sql/
   ```


Example output:
```json
{"num_statements": 1,
 "execution": {
   "retries":0,
   "txn_results": [
{
  "statement": 1,
  "tag": "SELECT",
  "start":"2022-04-08T15:55:30.659257101Z",
  "columns": [{"name":"v", "type":"STRING", "oid":25}],
  "rows": [
   {"v":"..."},
   {"v":"..."},
   {"v":"..."}
  ],
  "end":"2022-04-08T15:55:30.665283699Z"
}
  ]
}}
```

This currently executes using the "internal executor" and thus
does not support SQL sessions and transaction control. Some consequences:

- SET statements are ineffective.
- BEGIN/COMMIT/ABORT/SAVEPOINT etc are ineffective.

Security note: We place this logic under `/api/v2` and not
`/_admin/v1` because the latter uses session cookies and is thus
susceptible to [CSRF](https://en.wikipedia.org/wiki/Cross-site_request_forgery).
The `/api/v2` tree requires an explicit header and thus blocks CSRF.

Release note (api change): A new `/api/v2/sql/` endpoints enable
execution of simple SQL queries over HTTP.

---

Fixes https://github.com/cockroachdb/cockroach/issues/84385.

Release note (bug fix): The SQL execution HTTP endpoint did not
properly support queries with multiple result values. This is now
fixed. This bug was introduced when the feature was first implemented.

---

Release justification: this is a high-priority need for SQL Observability and makes it easier to port over functionality in DB Console. Furthermore, this endpoint is purely additive and can only be opted into by new frontend code.

knz committed:
server: new HTTP API to execute SQL statements

This commit introduces a new API endpoint `/api/v2/sql/`,
which can execute arbitrary SQL statements over HTTP.

The endpoint is authenticated, i.e. the client must provide
an appropriate `X-Cockroach-API-Session:` header on each request.

There is a Swagger spec to enable auto-documentation of the API.

The request payload should be submitted using JSON.
Summary of request fields:
- `database`: the current database, defaults to `defaultdb`
- `application_name`
- `timeout`: max execution time, defaults to 5s
- `max_results_size`: max size of results, defaults to ~10KB.
- `statements`: which statements to run. An array of {"sql",
  "arguments"} objects. The number of items in "arguments"
  must match the number of placehoders in "sql".
- `execute`: must be set to true to actually execute the SQL.
  Otherwise, the request is merely parsed to verify the syntax is correct.

Example use:

1. Generate some test data and create a user with access to it:

   ```
   $ cockroach gen example-data intro | cockroach sql --certs-dir=...
   $ cockroach sql --certs-dir=... \
       -e 'create user myuser with password abc'
       -e 'grant select on table intro.mytable to myuser'
   ```

2. Generate an HTTP API key, for example using the CLI (an HTTP
   request to /api/v2/login is also possible):

   ```
   $ cockroach session-auth login myuser  --certs-dir=certs
   ```

   (this prints out the session key, assign it manually to the $APIKEY env var.)

3. Run some HTTP API queries from the shell using `curl`. For example:

   ```
   $ curl -k -H "X-Cockroach-API-Session: $APIKEY" \
       -H "Content-Type: application/json" \
       --data-binary '{"database":"intro",
       "statements":[{"sql":"SELECT v FROM mytable WHERE l%2 = 0"}],
       "execute": true}' \
       https://127.0.0.1:8080/api/v2/sql/
   ```

Example output:
```json
{"num_statements": 1,
 "execution": {
   "retries":0,
   "txn_results": [
{
  "statement": 1,
  "tag": "SELECT",
  "start":"2022-04-08T15:55:30.659257101Z",
  "columns": [{"name":"v", "type":"STRING", "oid":25}],
  "rows": [
   {"v":"..."},
   {"v":"..."},
   {"v":"..."}
  ],
  "end":"2022-04-08T15:55:30.665283699Z"
}
  ]
}}
```

This currently executes using the "internal executor" and thus
does not support SQL sessions and transaction control. Some consequences:

- SET statements are ineffective.
- BEGIN/COMMIT/ABORT/SAVEPOINT etc are ineffective.

Security note: We place this logic under `/api/v2` and not
`/_admin/v1` because the latter uses session cookies and is thus
susceptible to [CSRF](https://en.wikipedia.org/wiki/Cross-site_request_forgery).
The `/api/v2` tree requires an explicit header and thus blocks CSRF.

Release note (api change): A new `/api/v2/sql/` endpoints enable
execution of simple SQL queries over HTTP.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84374)
<!-- Reviewable:end -->

knz committed:
server: fix the SQL execution endpoint

Release note (bug fix): The SQL execution HTTP endpoint did not
properly support queries with multiple result values. This is now
fixed. This bug was introduced when the feature was first implemented.

knz commented:
you'll need to backport #84386 too.

maryliag reviewed (approved):
As long as tests are passing and the necessary other PRs are also backported: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84374#-:-N6wv3l97NPC7MbMYzi4:b-4i2fv5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @abarganier and @xinhaoz)



<!-- Sent from Reviewable.io -->

dhartunian commented:
Auth code backport will be merged right after this one: https://github.com/cockroachdb/cockroach/pull/85553
